.changeset/moody-suns-smell.md

```diff
@@ -0,0 +1,37 @@
+---
+'@backstage/integration': minor
+'@backstage/integration-react': minor
+---
+
+Split `bitbucket` integration into `bitbucketCloud` and `bitbucketServer`
+(backwards compatible).
+
+In order to migrate to the new integration configs,
+move your configs from `integrations.bitbucket`
+to `integrations.bitbucketCloud` or `integrations.bitbucketServer`.
+
+Migration example:
+
+**Before:**
+
+```yaml
+integrations:
+  bitbucket:
+    - host: bitbucket.org
+      username: bitbucket_user
+      appPassword: app-password
+    - host: bitbucket-server.company.com
+      token: my-token
+```
+
+**After:**
+
+```yaml
+integrations:
+  bitbucketCloud:
+    - username: bitbucket_user
+      appPassword: app-password
+  bitbucketServer:
+    - host: bitbucket-server.company.com
+      token: my-token
```

Comment:
Wanna bounce some ideas around this to make sure we're not complicating things unnecessarily. This is ofc a pretty big change so wanna tread carefully.

Am I right in assuming that Bitbucket Cloud is only available on `bitbucket.org`? Or can you have an on-prem deployment of that too?

The main concern is whether the config split is necessary or if it should be done in a slightly different way. The three other approaches I can think of are:

```yaml
integrations:
  bitbucket:
    - host: bitbucket.org # Implicit type=cloud because bitbucket.org
      username: bitbucket_user
      appPassword: app-password
    - host: bitbucket-server.company.com
      token: my-token
```

```yaml
integrations:
  bitbucket:
    - host: bitbucket.org
      type: cloud # explicit type
      username: bitbucket_user
      appPassword: app-password
    - host: bitbucket-server.company.com
      type: server
      token: my-token
```

```yaml
integrations:
  bitbucket:
    - username: bitbucket_user
      appPassword: app-password
  bitbucketServer: # leave one of them as is and let the other use the separate integration type
    - host: bitbucket-server.company.com
      token: my-token
```

With all of there we'd have a pretty much complete separation of implementation with early type checks everywhere and adding a `type` to the Bitbucket config object.

Gotta say that looking at the problem it really seems sane to have them be completely separated like you suggest though, weird and unfortunate but perhaps necessary.

As a bit of a separate note, are the capabilities of the APIs very different as well? I'm thinking for things like the scaffolder action, whether it makes sense to split or abstract there. I'd say that would largely depend on whether the options are completely different between the two.

Reply:
Yes, bitbucket.org is the only host for Bitbucket Cloud.  The setup is a bit complicated with Bitbucket Cloud, Bitbucket Server (on-prem) and Bitbucket (Server) Data Center (managed Bitbucket Server). Bitbucket Server was formally known as Stash and was just rebranded after Atlassian bought Bitbucket (Cloud).

Reply:
In general, I like `bitbucket` + `bitbucketServer`, too, but makes it a bit hard for the migration and due to it being used at so many places.

Using `bitbucketCloud` + `bitbuckerServer` seemed easier for splitting it and more explicit. Unfortunately, it is a bit set up for confusion..

Theoretically, we could stick to `integrations.bitbucket` only, but that might be misaligned with other parts of the configuration.

There is an auth provider for Bitbucket Cloud using `bitbucket` though. But also here, many people asked whether this would support Bitbucket Server as well -- even though there is a warning at the documentation about it.

Reply:
Re API: yes, they are different and the reason why there were so many "isBitbucketCloud" checks at the code.

I've split the client at PR #10515

Reply:
overall, I think it would be great to extract all of this into separate modules for bitbucket cloud and server incl. UrlReader, etc. So far, this is not yet possible though.